xtask/src/set_cairo_version.rs

```diff
@@ -45,83 +46,193 @@ pub fn main(args: Args) -> Result<()> {
-        let mut new_dep = InlineTable::new();
+        // In `[dependencies]` and `[dev-dependencies]` tables, always use crates.io requirements
+        // so that downstream (Scarb) can reliably patch them with the `[patch.crates-io]` table.
+        let mut new_dep = InlineTable::from_iter([("version", match &args.spec.version {
+            Some(version) => Value::from(version.to_string()),
+            None => Value::from("*"),
+        })]);
 
-        if let Some(version) = &args.spec.version {
-            new_dep.insert("version", version.to_string().into());
-        }
+        copy_dependency_features(&mut new_dep, dep);
 
-        // Add a Git branch or revision reference if requested.
-        if args.spec.rev.is_some() || args.spec.branch.is_some() {
-            new_dep.insert("git", "https://github.com/starkware-libs/cairo".into());
-        }
+        *dep = new_dep.into();
+        simplify_dependency_table(dep)
+    }
 
-        if let Some(branch) = &args.spec.branch {
-            new_dep.insert("branch", branch.as_str().into());
-        }
+    deps.fmt();
+    deps.sort_values();
+
+    eprintln!("[{table}]");
+    for (key, dep) in deps.iter().filter(|(key, _)| is_cairo_crate(key)) {
+        eprintln!("{key} = {dep}");
+    }
+}
+
+fn edit_patch(cargo_toml: &mut DocumentMut, args: &Args) {
+    let patch = cargo_toml["patch"].as_table_mut().unwrap()["crates-io"].as_table_mut().unwrap();
+
+    // Clear any existing entries for Cairo crates.
+    patch.retain(|key, _| !is_cairo_crate(key));
+
+    // Leave this section empty if we are requested to just use a specific version.
+    if args.spec.rev.is_some() || args.spec.branch.is_some() || args.spec.path.is_some() {
+        // Patch all Cairo crates that exist, even if this project does not directly depend on them,
+        // to avoid any duplicates in transient dependencies.
+        for &dep_name in CAIRO_CRATES {
+            let mut dep = InlineTable::new();
+
+            // Add a Git branch or revision reference if requested.
+            if args.spec.rev.is_some() || args.spec.branch.is_some() {
+                dep.insert("git", "https://github.com/starkware-libs/cairo".into());
+            }
+
+            if let Some(branch) = &args.spec.branch {
+                dep.insert("branch", branch.as_str().into());
+            }
+
+            if let Some(rev) = &args.spec.rev {
+                dep.insert("rev", rev.as_str().into());
+            }
+
+            // Add local path reference if requested.
+            // For local path sources, Cargo is not looking for crates recursively therefore, we
+            // need to manually provide full paths to Cairo workspace member crates.
+            if let Some(path) = &args.spec.path {
+                dep.insert(
+                    "path",
+                    path.join("crates").join(dep_name).to_string_lossy().into_owned().into(),
+                );
+            }
```

Comment:
How about checking the path for validity? At the moment, we can pass an arbitrary string which will be written to the manifest before calling `cargo fetch`

Reply:
This is fine for me because the aforementioned `cargo fetch` will always be called and thus validate it.

Same thing applies to version numbers and git refs.